### DISCUSSION TITLE
[EDITOR] Features/suggestions thread

### DISCUSSION THREAD

**Discussion by JoseVarelaP** [TeamRizu/OutFox#328] [03 - Feature Requests]
This post is to have a centralized area for potential ideas for features for the Editor, given the constant updates on said element for the upcoming Alpha 5, which already has some significant changes under the hood, while having to deal with the legacy code it had before.

## Elements already done
- [X] Two Player support (planning on multi-player once extra players are conceptualized)
- [X] Complete Mod support (Already included fail-safes for screen manipulation resets as well)
[Mods Showcase](https://josevarela.xyz/img/EditorMedia/mods.mp4)
- [X] Added Go To Beat
- [X] Preview mode (Rather than a full playback mode, it only plays from the same charting screen, to allow quick previews of the chart playing for analyzing).
[Preview Mode Showcase](https://josevarela.xyz/img/EditorMedia/preview.mp4)
- [X] OS-styled Copy/Paste
[Copy/Paste Showcase](https://josevarela.xyz/img/EditorMedia/copy_paste.mp4)
- [X] MacOS native ⌘ (CMD) key support
- [X] Dynamic Undo/Redo (No more memory swapping from legacy).
- [X] Toggling of measure lines on playback mode
- [x] Range-Based note deletion (Select an area of notes and just press delete to erase that section, rather than eating a beat)
[Preview Mode Showcase](https://josevarela.xyz/img/EditorMedia/deletesection.mp4)
- [X] Note stutter generation (With/Without mines option for charters)
[Note Stutter Showcase](https://josevarela.xyz/img/EditorMedia/stutter.mp4)
- [X] Chart2Lua (Whole/Segment of chart)
- [X] Create a module styled update methodology for themers, allowing for even more control to customise the editor ***(Need feedback about this approach)***.
- [X] Dynamic Editor NoteSkin switching.
[Noteskin switching Showcase](https://josevarela.xyz/img/EditorMedia/noteskin.mp4)
- [X] Lua based help menu (Original implementation was a ScreenMiniMenu, back from 3.9, which is extremely limited).
- [X] Highlight on Routine notes upon player selection (Highlights the notes of the player currently editing, while the other player's notes will become semi-transparent)
[Player switching Showcase](https://josevarela.xyz/img/EditorMedia/routineplayers.mp4)
- [X] BothAtOnce support
- [X] Make TextEntry not listen to GAME_BUTTON input for text entry (It's support to be for text, why was it mapped to Game buttons!?)
- [X] ModList of active mods per player in Playback mode.
- [X] Theme support for modifying labels
- [X] Dynamic Snap icons (They can be Actorframes, models, whatever)
- [X] XMod -> CMod (and viceversa) scrolling.
- [X] Warp / Stop visual area highlight on Notefield.
[Stops Showcase](https://josevarela.xyz/img/EditorMedia/stops.mp4)
[Warps Showcase](https://josevarela.xyz/img/EditorMedia/warps.mp4)
- [X] Support for changing Time Signatures on Song Timing.
[Time Signature Showcase](https://josevarela.xyz/img/EditorMedia/timesignatures.mp4)
- [X] MouseWheel note scrolling.
- [X] Window Title updating based on Song Title (Even shows native titles)
- [X] Redo/Undo stores full note/timing history
- [X] Previewing Note Flashing
- [X] Notification system (This is a replacement of using SystemMessages to allow showing more than one at a time).
- [X] FG Changes prompt for adding files like BGChanges.
- [X] Support for Minefields and LiftHolds
- [X] Lua functions to set the Song/Step timing
- [X] Lua functions to set the Snap Mode Type for arrow placement
- [X] Lua function to set a marker selection area
- [X] Marker selection for Attacks (Select an area, edit an existing attack and asign the selection to be the attack's start and end).
- [x] Support for new note types.

## Considered features

- [ ] Mouse Support (Initial version already has support for selecting an area of notes, but not individual ones, given how NoteField has been implemented previously).
- [ ] More visual/backend indication of Song/Step timing.
- [ ] Light mode chart editing
- [ ] Cursor Autoadvancing (An editor option, already has a prototype which needs some fixing, place a note, and the cursor will automatically move to the next quantization currently selected, so after placing a 4th note, it will go to the next spot for a 4th note).

Any possible suggestions, please post them in this discussion for them to be considered.
- Jose_Varela

**Comment by snil4** [comment]
It seems like you plan to implement a lot of what I wanted to the editor, but here are a few suggestions from using arrow vortex a lot and slightly going to outfox for small changes:
- new keymapings to go together with mouse support
- quicker way to do things like scrolling, play testing, and enabling/disabling assist sounds
- scrolling with the scroll wheel
- waveforms in the editor for easier syncing
- less menus and easier navigation
- recording charts (adding notes while the song is playing)
- dynamic window reisizing (maybe more of outfox problem but could be a nice feature)
- dragging the notefield with the mouse
- "save as" to select save format (.sm or .ssc)

If anything is already in outfox or planed don't mind me, but nonetheless all those features are important for me when using a chart editor.

**Reply by moruzerinho6** [reply]
Why would we give the option to save as .sm

**Reply by snil4** [reply]
sm 3.95, nITG, OpenITG, and what about other formats? saving as .chart, .ucs, .osu (osumania), .bms, and many more seem useful enough for an editor that wants to cover so many types of games.

**Reply by lomdr** [reply]
3.9+ is still getting some love community side as well, with there being a DDR A20+ theme being made for it.  3.9+ also doesn't support SSC files

**Comment by PachiSystems** [comment]
I would love a way to be able to edit the light patterns for my cab along with the steps. Or have a better autogen function for lights. At the moment they all just flash with each step unless I use an external program to generate a light chart based on one of the stepcharts present.

**Comment by RyutoSetsujin** [comment]
I don't know if this has been considered, but I'd like a "play display" option so you can see how your charts could be played (similar to ArrowVortex's Dancing Bot) on a real controller.

**Comment by luizsan** [comment]
Lua bindings for:
- Adding, removing, reading and writing Timing Data on current beat or range of beats
- [done] ~~Changing between Song/Step Timing states~~
- Changing between other Steps from the same Song
- [done] ~~Changing current snap, note type~~
- [done] ~~Selecting beats~~

Basically, the more we can control the editor by using Lua the same way we can control it using the built-in menus, the better.

For features that I'd like to have but might be a huge task to implement:
- More undo levels
- Waveform visualization

**Reply by JoseVarelaP** [reply]
A couple questions given the suggestions.

Given the snap form, I would assume something in the form of choosing via a enum method?
For choosing beats, would it be sometihng like a two argument function to define the start and ending beat?

**Reply by luizsan** [reply]
For the snap function: passing an int value or enum string as the function argument would work just fine.

For choosing beats: yeah, that also works! Preferably, the function would automatically choose the math.min of the arguments for the selection start, and the math.max for the selection end (or length? I don't know what would be better in this case), and having only one value (or two identical values) would set the selection to a single beat.

**Reply by JoseVarelaP** [reply]
> Preferably, the function would automatically choose the math.min of the arguments for the selection start, and the math.max for the selection end (or length? I don't know what would be better in this case)

The selection being the absolute value (start and end) would be easier to calculate.

**Comment by makoflagkk** [comment]
A new feature I would like to see come to Outfox Alpha 5 is adding a cheering flag to the game engine. This would adding cheering if the player reached a certain combo. You would add the crowd cheering and the announcer say something at the same time. I just don't want it for every song but rather to be added to a theme.

**Reply by JoseVarelaP** [reply]
The game already does this internally every 100 combos. There's no need to add this to the editor.

The behavior that you want can also be done by the song itself with some Lua built in with FGChanges.
```lua
local alreadyAchievedCombo = false
local myComboThreshold = 220
ScoreChangedMessageCommand=function(self,params)
    -- Check if the player has already achieved the desired goal.
    if alreadyAchievedCombo then return end
    -- For this example, let's say we only want to play this with player 1.
    if params.PlayerNumber ~= PLAYER_1 then return end

    local stats = STATSMAN:GetCurStageStats():GetPlayerStageStats(params.PlayerNumber)
    -- Get the player's current combo.
    local combo = stats:GetCurrentCombo()

    if combo >= myComboThreshold then
        -- Call the announcer line.
        SOUND:PlayAnnouncer("my custom combo line")
        alreadyAchievedCombo = true
    end
end
```
